### ISSUE TITLE
Refactor Debugging Protocol

### ISSUE THREAD

**Issue by molant** [webhintio/hint#1398]
`utils-debugging-protocol` needs some ❤
The interaction with the debugging protocol is a bit complex and ~~we~~I've kept adding hacks during the last few months making the code messiers and messier. It's time to pay the debt.

Some examples of issues we have or had recently:

1. #1395 
2. https://github.com/webhintio/rfcs/issues/44
3. #1325
4. Also `rawResponse` can be empty in most cases (similar to #1395 but affecting `performance-budget`)
5. #1403

**Comment by molant** [comment]
I've already started with this refactor while trying to fix `4` and have found something interesting:
It looks like Chrome will do 2 requests if it gets a 404, at least when testing localhost. I assume there're some other heuristics behind.

For this HTML (`script4.js` and `script6.js` return a 404):

```html
<!doctype html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>Common connector test page</title>
        <script>
            var a = 'a';
        </script>
        <script src="/script3.js"></script>
        <script src="/script4.js"></script>
        <script src="/script5.js"></script>
        <style>
            h1 {
                font-size: 2em;
            }
        </style>
        <link rel="stylesheet" href="./style.css"/>
    </head>

    <body>
        <h1>Common connector test page</h1>
        <p>The content of this HTML is used to validate that all connectors that parse full HTML send the same events.</p>
        <p>These two images are to test that we are not entering in a infinite loop when we have more than one element with the same src or href</p>
        <img src='http://localhost/nellie.png'>
        <img src='http://localhost/nellie.png'>
    </body>
</html>
```

We get the following network trace:
![image](https://user-images.githubusercontent.com/606594/46965270-39809a00-d05f-11e8-87d0-d428a73a4658.png)

I have no idea of how we were passing the tests before (yet another issue) but I don't think we should hide when the browser does another requests and we should send a `fetch::start` and `fetch::error` for each request.

@webhintio/contributors what do you think?

**Comment by sarvaje** [comment]
IMO, if the browser send two request, we should emit the events two times. In case some "report" is duplicated, is the `report` logic the one should remove duplicate messages.

**Comment by antross** [comment]
I agree with emitting the events two times.

FWIW, the double-download is likely a result of [speculative downloading](https://developer.mozilla.org/en-US/docs/Web/HTML/Optimizing_your_pages_for_speculative_parsing) by the HTML parser. Basically once blocked on an earlier script element, browsers will scan ahead for other script, style, and image tags and start fetching them. However, since the resolved URL can change between speculative parsing and actual parsing (e.g. a `<base>` element was dynamically added) the resource may need to be fetched again. Chrome likely automatically ignores 404s and tries again during the actual parse instead of checking to see if the resolved URL actually changed.

**Comment by molant** [comment]
There's a PR for this now #1409
